### ISSUE TITLE
Multiplayer Desync Issue

### ISSUE THREAD

**Issue by ubergarm** [hooap/SatisfiedStorage#4]
I love `SatisfiedStorage` in solo play, but I tested it for my multiplayer modpack [Happy Accidents](https://steamcommunity.com/sharedfiles/filedetails/?id=2257918295) and it causes a desync when the pawn goes to re-fill the stack.

To recreate:
1. Install Multiplayer Mod (or use above modpack and config) and SatisfiedStorage
2. Host a game w/ the Arbiter enabled (so you can detect desyncs alone w/o bothering your friends to test)
3. Setup a 1x1 stockpile for just simple meals for example
4. Let your pawns fill up the stockpile
5. Set the restock percentage to 80%
6. When the stack drops below the threshold and a pawn gets the job to restock it, the game desyncs.

My guess is that the restock percentage isn't being synchronized. No idea if looking at this patch that adds multiplayer api hooks is useful in this situation [recent example of syncing information for multiplayer mod in allow tool](https://github.com/notfood/RimWorld-AllowTool/commit/6b197b8d180d8827d101d9e16bee3d1f58367461)

The last entry in the `player.log` before the desync is as follows:
```
Invoked Boolean TryTakeOrderedJob(Verse.AI.Job, Verse.AI.JobTag) on Verse.AI.Pawn_JobTracker with 2 params TakeInventory (Job_-1) A=Thing_MealSimple374451, 0
```

It doesn't look like anyone at [Multiplayer Compatibility Mod](https://github.com/rwmt/Multiplayer-Compatibility) is patching it as of yet.

Cheers!

**Comment by ubergarm** [comment]
I found two more possibly useful references, `LWM Deep Storage` mod is currently getting a patch to multiplayer sync their "Eject" button and had some discussion in the Multiplayer Compat Mod discord channel `#dev-chat` here: https://discord.gg/8ttTft5

Here is the wiki to their multiplayer API which from what I understand allows you to trigger it to sync the gui updates:
https://github.com/rwmt/MultiplayerAPI/wiki

No pressure to look into it, just gathering all the info in once place! Take care!

**Comment by ubergarm** [comment]
I'm working on a [solution over here](https://github.com/hooap/SatisfiedStorage/compare/master...ubergarm:multiplayer-patch), haven't opened a PR yet as this is my first MP patch. Hoping to get some input from the [multiplayer discord](https://discordapp.com/channels/524286515644203028/568070216626470922/773257728948305930) first before submitting to ya!

**Comment by ubergarm** [comment]
I tried everything I could think of, but still get a multiplayer desync for the corner case where a pawn is currently hauling to the stockpile when someone slides the hysteresis fill percent down lower thus canceling the job.

From my testing and forcing all pawns to stop jobs when the slider is touched, it seems like somehow the AI is giving pawns different new jobs on different machines causing the desync, but tbh i don't really know and this was my first attempt at MP patching...

I'm gonna punt on this patch for now, maybe later someone will figure it out.. lol cheers!

**Comment by ubergarm** [comment]
Well, after a lot of work and testing it seems like using a complex object like `StorageSettings` as the key in a dictionary doesn't sync deterministically for mutiplayer. I don't fully understand it, but seems like the key is using either a local memory ref or some per process startup randomized `GetHashCode()`..

So I combined aspects of your mod with aspects of `Variety Stockpile Matters` and created a new mod [KanbanStockpile](https://steamcommunity.com/sharedfiles/filedetails/?id=2287142613). I put you and other folks in credits, thanks much and cheers!

**Comment by burak-aydin** [comment]
yeah i spent like 5-6 hours trying to sync settings for multiplayer as well with no luck. it would probably require to rewrite how it holds values. Inital aim of this mod was to port it to 1.2 and i dont have much time playing/modding rimworld due to real life right now. sorry about replying late,  also your mod looks great.

**Comment by ubergarm** [comment]
Yeah, I tried a lot of ways to sync those stockpile settings too, and got it *kind of* working.  

But once I realized it may need a different data structure to hold the values, I decided to try a rewrite incorporating the best of both your `Satisfied Storage` and a little of `Variety Matters Stockpile`.

Thanks so much for putting your stuff up on github and sharing! It's been a fun project for me while I'm hiding out in my apartment not working right now, haha...

Take care and hope you get some time to play whatever you're into these days.

Cheers!
